apps/application-system/api/src/app/modules/application/application.controller.ts

```diff
@@ -639,70 +643,90 @@ export class ApplicationController {
 
     const templateId = existingApplication.typeId as ApplicationTypes
     const template = await getApplicationTemplateByTypeId(templateId)
+    return withCodeOwner(template.codeOwner, async () => {
+      const helper = new ApplicationTemplateHelper(
+        existingApplication as BaseApplication,
+        template,
+      )
 
-    const helper = new ApplicationTemplateHelper(
-      existingApplication as BaseApplication,
-      template,
-    )
+      const userRole = template.mapUserToRole(
+        user.nationalId,
+        existingApplication as BaseApplication,
+      )
 
-    const userRole = template.mapUserToRole(
-      user.nationalId,
-      existingApplication as BaseApplication,
-    )
+      const providersFromRole = userRole
+        ? helper.getApisFromRoleInState(userRole)
+        : []
+
+      const namespaces = await getApplicationTranslationNamespaces(
+        existingApplication as BaseApplication,
+      )
+      const intl = await this.intlService.useIntl(namespaces, locale)
 
-    const providersFromRole = userRole
-      ? helper.getApisFromRoleInState(userRole)
-      : []
+      const templateApis: TemplateApi[] = []
 
-    const namespaces = await getApplicationTranslationNamespaces(
-      existingApplication as BaseApplication,
-    )
-    const intl = await this.intlService.useIntl(namespaces, locale)
+      for (let i = 0; i < externalDataDto.dataProviders.length; i++) {
+        const found = providersFromRole.find(
+          (x) => x.actionId === externalDataDto.dataProviders[i].actionId,
+        )
 
-    const templateApis: TemplateApi[] = []
+        if (found) {
+          templateApis.push(found)
+        } else {
+          throw new BadRequestException(
+            `Current user is not permitted to update external data in this state with actionId: ${externalDataDto.dataProviders[i].actionId}`,
+          )
+        }
+      }
 
-    for (let i = 0; i < externalDataDto.dataProviders.length; i++) {
-      const found = providersFromRole.find(
-        (x) => x.actionId === externalDataDto.dataProviders[i].actionId,
+      await this.validationService.validateIncomingExternalDataProviders(
+        existingApplication as BaseApplication,
+        templateApis,
+        user.nationalId,
       )
 
-      if (found) {
-        templateApis.push(found)
-      } else {
-        throw new BadRequestException(
-          `Current user is not permitted to update external data in this state with actionId: ${externalDataDto.dataProviders[i].actionId}`,
+      const updatedApplication = await this.templateApiActionRunner.run(
```

Comment:
Would be ideal to wrap this with logging context, as it may invoke all kinds of things.